### PR TITLE
Timescale support

### PR DESCRIPTION
**Pull request by mayhem**
This PR removes all traces of influx (\ø/) and adds support for timescale. This PR is rather long, but generally not too difficult to digest. Besides adding support for timescale it also creates the necessary infrastructure for having a test instance of ListenBrainz.

While this PR should be close to being merged, DO NOT MERGE this PR. The timing of merging this needs to be carefully considered as per our migration instructions.

The code in this PR is love at https://test.listenbrainz.org .

### CODE REVIEW

File: admin/timescale/test.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import json
```
Comment (paramsingh) [inline admin/timescale/test.py:1]: Is this file needed?

File: listenbrainz/listenstore/listenstore.py [hunk 2]
```diff
@@ -63,11 +58,11 @@ def import_listens_dump(self, archive_path, threads=None):
         """
         raise NotImplementedError()
 
-
-    def fetch_listens(self, user_name, from_ts=None, to_ts=None, limit=DEFAULT_LISTENS_PER_FETCH):
+    def fetch_listens(self, user_name, from_ts=None, to_ts=None, limit=DEFAULT_LISTENS_PER_FETCH, time_range=None):
         """ Check from_ts, to_ts, and limit for fetching listens
             and set them to default values if not given.
         """
+        self.log.info("fetch listens called")
```
Comment (paramsingh) [inline listenbrainz/listenstore/listenstore.py:65]: Do we need this log?

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 3]
```diff
@@ -68,27 +67,35 @@ def get_listen_count_for_user(self, user_name, need_exact=False):
...
             count = cache.get(user_key, decode=False)
             if count:
                 return int(count)
 
-        try:
-            results = self.influx.query('SELECT count(*) FROM ' + get_escaped_measurement_name(user_name))
-        except (InfluxDBServerError, InfluxDBClientError) as e:
-            self.log.error("Cannot query influx: %s" % str(e), exc_info=True)
-            raise
+        if "PYTEST_CURRENT_TEST" in os.environ:
```
[the remark this answers is not in the document]
Comment (paramsingh) [reply listenbrainz/listenstore/timescale_listenstore.py:75]: Because it doesn't really make sense to add test specific behaviour. We're not testing what'll actually be run, why bother? Maybe we should just remove the test.

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 4]
```diff
@@ -179,177 +172,118 @@ def get_timestamps_for_user(self, user_name):
...
 
-        # Quote single quote characters which could be used to mount an injection attack.
-        # Sadly, influxdb does not provide a means to do this in the client library
-        query = 'SELECT * FROM ' + get_escaped_measurement_name(user_name)
+        if not time_range:
+            time_range = 3
+        if time_range < 0:
+            max_timestamp_window = -1
+        else:
+            max_timestamp_window = 432000 * time_range
```
Comment (paramsingh) [inline listenbrainz/listenstore/timescale_listenstore.py:251]: Could we move this magic number into a constant

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 5]
```diff
@@ -179,177 +172,118 @@ def get_timestamps_for_user(self, user_name):
...
 
         if from_ts is not None:
-            query += "WHERE time > " + get_influx_query_timestamp(from_ts)
+            to_ts = from_ts + max_timestamp_window
         else:
-            query += "WHERE time < " + get_influx_query_timestamp(to_ts)
+            from_ts = to_ts - max_timestamp_window
+            if "PYTEST_CURRENT_TEST" in os.environ:
```
Comment (paramsingh) [inline listenbrainz/listenstore/timescale_listenstore.py:257]: again, I seriously think we should fix the tests instead of adding test specific behaviour. tests won't reflect actual behaviour and it'll be hard to understand why this code was added, otherwise.

File: listenbrainz/webserver/static/js/src/RecentListens.test.tsx [hunk 6]
```diff
@@ -28,6 +28,7 @@ const {
   spotify,
   user,
   webSocketsServerUrl,
+  tryHarder,
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/RecentListens.test.tsx:31]: What does this do? Perhaps we can use a more descriptive name. Right now, I have no idea what try harder means.

File: admin/sql/create_tables.sql [hunk 7]
```diff
@@ -142,4 +142,6 @@ CREATE TABLE recording_feedback (
     created                 TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW()
 );
 
+GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO listenbrainz;
```
[the remark this answers is not in the document]
Comment (mayhem) [reply admin/sql/create_tables.sql:145]: Its a bodge that outgrew from a DB setup issue - we're working on resolving that.

File: listenbrainz/listenstore/timescale_listenstore.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/listenstore/timescale_listenstore.py:75]: Fortunately this is no longer needed (thanks timescale, not Obama!) as continuous aggregates have been improved to always give a correct result (at the expense of a bit of performance).

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 8]
```diff
@@ -98,79 +105,67 @@ def reset_listen_count(self, user_name):
         """
         self.get_listen_count_for_user(user_name, need_exact=True)
 
-    def _select_single_value(self, query, measurement):
-        try:
-            results = self.influx.query(query)
-        except Exception as err:
-            self.log.error("Cannot query influx: %s" % str(err), exc_info=True)
-            raise
+    def _select_single_timestamp(self, select_min_timestamp, user_name):
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/listenstore/timescale_listenstore.py:108]: fixed.

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 9]
```diff
@@ -179,177 +174,124 @@ def get_timestamps_for_user(self, user_name):
...
 
         submit = []
         user_names = {}
         for listen in listens:
             user_names[listen.user_name] = 1
-            submit.append(listen.to_influx(quote(listen.user_name)))
+            submit.append((listen.ts_since_epoch, listen.recording_msid,
+                           listen.user_name, listen.to_timescale()))
+
+        query = """INSERT INTO listen (listened_at, recording_msid, user_name, data)
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/listenstore/timescale_listenstore.py]: fixed.

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 10]
```diff
@@ -179,177 +174,124 @@ def get_timestamps_for_user(self, user_name):
...
+                    VALUES %s
+                    ON CONFLICT (listened_at, recording_msid, user_name)
+                        DO NOTHING
+                        RETURNING listened_at, recording_msid, user_name"""
+
+        inserted_rows = []
+        conn = timescale.engine.raw_connection()
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/listenstore/timescale_listenstore.py:214]: Yes, because we're using the execute_values method for psycopg2 which allows for more efficient large INSERTs. I think SQLA has support for it, but rather than drinking more SQLA kool-aid, I opted to stick closer to psycopg2.

File: admin/timescale/test.py (hunk 1 above)
Comment (mayhem) [reply admin/timescale/test.py:1]: Nope!

File: admin/sql/create_tables.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (shivam-kapila) [reply admin/sql/create_tables.sql:145]: The GRANT statements have been removed now

File: listenbrainz/listenstore/timescale_listenstore.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shivam-kapila) [reply listenbrainz/listenstore/timescale_listenstore.py:75]: Fixed

File: listenbrainz/webserver/static/js/src/RecentListens.test.tsx (hunk 6 above)
Comment (shivam-kapila) [reply listenbrainz/webserver/static/js/src/RecentListens.test.tsx:31]: Updated to `searchLargerTimeRange`

File: listenbrainz/listenstore/listenstore.py (hunk 2 above)
Comment (mayhem) [reply listenbrainz/listenstore/listenstore.py]: Removed.
